solarforecastarbiter/metrics/preprocessing.py

```diff
@@ -895,3 +934,89 @@ def _name_pfxobs(current_names, forecast, i=1):
         return _name_pfxobs(current_names, new_name, i + 1)
     else:
         return forecast_name
+
+
+def get_outage_periods(forecast, start, end, outages):
+    """Converts report outage periods to forecast data periods to
+    drop from analysis.
+
+    Parameters
+    ----------
+    forecast: Forecast
+    start: pd.Timestamp
+    end: pd.Timestamp
+    outages list of ReportOutages
+
+    Returns
+    -------
+    list of dicts
+        List of dicts with start and end keys. Times between these values
+        should not be included in analysis.
+    """
+    # First, determine a list of forecast issue times that include data that
+    # falls within the report
+
+    # Get total forecast horizon to know how far back to look for issue times
+    total_forecast_horizon = forecast.lead_time_to_start + forecast.run_length
+
+    # Look back another day, so we can start from a day's first issue time
+    # without undershooting the forecast horizon
+    lookback_period = total_forecast_horizon + pd.Timedelta('1D')
+
+    # Convert start to UTC so we can replace with the UTC-defined issue time
+    # to get the first issue time inside the lookback period.
+    lookback_start = start.tz_convert('UTC') - lookback_period
+    issue_start = lookback_start.replace(
+        hour=forecast.issue_time_of_day.hour
+    )
+
+    # Get all possible issue times that contribute to the report
+    issue_times = pd.date_range(
+        issue_start,
+        end.tz_convert('UTC'),
+        freq=forecast.run_length
+    )
+
+    outage_periods = []
+    # For each outage, if a forecast submission/issue_time falls within
+    # the outage. Create start/end bounds for the forecast data to exclude.
+    for outage in outages:
+        outage_submissions = issue_times[
+            (issue_times >= outage.start) & (issue_times <= outage.end)
+        ]
```

Comment:
I understand the desire for fuzzy bounds here, but it seems like it could come back to bite us. I think if the buffer is user adjustable and defaulted to 0, it would be okay but I'm not sure where to store that since report parameters are static, and outages can be adjusted like values. But if we're asking a user to configure some window for exclusion we could also just say "Consider proximity to issue times when creating outages".

Reply:
Fair enough, but what are you thinking we'd do for the standard API outages? Add a buffer when defining them in the database? That seems not ideal also.

Reply:
I think that if you need to supply some wiggle room, you could overwrite the standard API outage with a report outage, or ignore the API outages entirely and provide your own. My motivation for the  `exclude_system_outages` report parameter, discussed in https://github.com/SolarArbiter/solarforecastarbiter-api/pull/325 was the idea that you could ignore the built in system outages if you weren't happy with how we use them, and I think this is one of those exceptions.

The report outages can be updated on the fly too, so If they get set and then we realize half the forecasts didn't make the first post-outage submission, the outages could be updated and the report recomputed, but the system outages will exist as a static record of when the system was actually down.